### PR TITLE
write file locations into html report

### PR DESCRIPTION
**Pull request by reinsch82**
Part of https://github.com/gradle/gradle-private/issues/4279

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

File: .teamcity/subprojects.json [hunk 1]
```diff
@@ -982,7 +982,7 @@
   {
     "name": "problems",
     "path": "platforms/ide/problems",
-    "unitTests": false,
+      "unitTests": true,
```
Comment (donat) [inline .teamcity/subprojects.json:985]: unnecessary whitespace

File: platforms/ide/problems/build.gradle.kts [hunk 2]
```diff
@@ -65,6 +65,10 @@ dependencies {
     implementation(projects.logging)
     implementation(projects.messaging)
 
+    testImplementation(projects.stdlibKotlinExtensions)
+
```
Comment (donat) [inline platforms/ide/problems/build.gradle.kts:69]:
```suggestion
```

File: platforms/ide/problems/build.gradle.kts [hunk 3]
```diff
@@ -65,6 +65,10 @@ dependencies {
     implementation(projects.logging)
     implementation(projects.messaging)
 
+    testImplementation(projects.stdlibKotlinExtensions)
```
Comment (donat) [inline platforms/ide/problems/build.gradle.kts:68]: Is this needed?

File: platforms/ide/problems/src/main/kotlin/org/gradle/problems/internal/impl/DefaultProblemsReportCreator.kt [hunk 4]
```diff
@@ -85,46 +86,69 @@ class DefaultProblemsReportCreator(
...
-                                property("path", path)
+        report.onProblem(JsonProblemWriter(problem, failureDecorator, failureFactory))
+    }
+}
+
+class JsonProblemWriter(private val problem: Problem, private val failureDecorator: FailureDecorator, private val failureFactory: FailureFactory) : JsonSource {
+    override fun writeToJson(jsonWriter: JsonWriter) {
+        with(jsonWriter) {
+            jsonObject {
+                if (problem.locations.isNotEmpty()) {
```
Comment (donat) [inline platforms/ide/problems/src/main/kotlin/org/gradle/problems/internal/impl/DefaultProblemsReportCreator.kt:97]: Shouldn't we filter for file locations here and check whether the result is empty?

File: platforms/ide/problems/src/test/kotlin/org/gradle/problems/internal/impl/JsonWriterTest.kt [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+import org.gradle.internal.extensions.stdlib.uncheckedCast
+import org.hamcrest.MatcherAssert.assertThat
+import org.hamcrest.Matchers.hasEntry
+import org.junit.Test
+import java.io.StringWriter
+
+class JsonWriterTest {
+
+    @Test
+    fun `complex structure is valid JSON`() {
```
Comment (donat) [inline platforms/ide/problems/src/test/kotlin/org/gradle/problems/internal/impl/JsonWriterTest.kt:32]: It's not clear to me what do we test here. A better method name or more descriptive sample test data would be useful.

File: platforms/ide/problems/build.gradle.kts (hunk 2 above)
Comment (reinsch82) [reply platforms/ide/problems/build.gradle.kts:69]: I wanted to separate project from library dependencies

File: platforms/ide/problems/build.gradle.kts (hunk 3 above)
Comment (reinsch82) [reply platforms/ide/problems/build.gradle.kts:68]: Yes, for the new JsonWriterTest class

**Comment by reinsch82** [comment]
@bot-gradle test this

**Review by donat — APPROVED** [review]
